pkg/service/storage/ucan.go

```diff
@@ -172,6 +179,100 @@ func NewUCANServer(storageService Service, options ...server.Option) (server.Ser
+					trnsfInv, err := replica.Transfer.Invoke(
+						storageService.ID(),
+						storageService.ID(),
+						storageService.ID().DID().GoString(),
+						replica.TransferCaveats{
+							Space:    cap.Nb().Space,
+							Blob:     cap.Nb().Blob,
+							Location: cap.Nb().Location,
+							Cause:    inv.Link(),
+						},
+					)
+					if err != nil {
+						return replica.AllocateOk{}, nil, failure.FromError(err)
+					}
+					for block, err := range inv.Blocks() {
+						if err != nil {
+							return replica.AllocateOk{}, nil, fmt.Errorf("iterating replica allocate invocation blocks: %w", err)
+						}
+						if err := trnsfInv.Attach(block); err != nil {
+							return replica.AllocateOk{}, nil, fmt.Errorf("failed to replica allocate invocation block (%s) to transfer invocation: %w", block.Link().String(), err)
+						}
+					}
+
+					// read the location claim from this invocation to obtain the DID of the URL
+					// to replicate from on the primary storage node.
+					br, err := blockstore.NewBlockReader(blockstore.WithBlocksIterator(inv.Blocks()))
+					if err != nil {
+						return replica.AllocateOk{}, nil, failure.FromError(err)
+					}
+					claim, err := delegation.NewDelegationView(cap.Nb().Location, br)
+					if err != nil {
+						return replica.AllocateOk{}, nil, failure.FromError(err)
+					}
+
+					// TODO since there is a slice of capabilities here we need to validate the 0th is the correct one
+					// unsure what `With()` should be compared with for a capability.
+					lc, err := assert.LocationCaveatsReader.Read(claim.Capabilities()[0].Nb())
+					if err != nil {
+						return replica.AllocateOk{}, nil, failure.FromError(err)
+					}
+
+					if len(lc.Location) < 1 {
+						return replica.AllocateOk{}, nil, failure.FromError(fmt.Errorf("location missing from location claim"))
+					}
+
+					// TODO: which one do we pick if > 1?
+					replicaAddress := lc.Location[0]
+
+					// FIXME: use a real context, requires changes to server
+					ctx := context.TODO()
+					allocateResp, err := storageService.Capabilities().BlobAllocate(ctx, &capabilities.BlobAllocateRequest{
+						Space: cap.Nb().Space,
+						Blob:  cap.Nb().Blob,
+						Cause: inv.Link(),
+					})
+					if err != nil {
+						return replica.AllocateOk{}, nil, failure.FromError(err)
+					}
+
+					// will run replication async, sending the receipt of the transfer invocation
+					// to the upload service.
+					if err := storageService.Replicator().Replicate(ctx, &replicator.Task{
+						Space:      cap.Nb().Space,
+						Blob:       cap.Nb().Blob,
+						Source:     replicaAddress,
+						Sink:       allocateResp.Address.URL,
+						Invocation: trnsfInv,
+					}); err != nil {
+						return replica.AllocateOk{}, nil, failure.FromError(fmt.Errorf("failed to enqueue replication task: %w", err))
+					}
+
+					return replica.AllocateOk{Size: allocateResp.Size}, fx.NewEffects(fx.WithFork(fx.FromInvocation(trnsfInv))), nil
```

Comment:
the RFC for this work states:
> The blob/replica/allocate task receipt includes an async task that will be performed by the storage node - blob/replica/transfer. The blob/replica/transfer task is completed when the storage node has transferred the blob from its location to the storage node
>...
> Client can poll the upload service for the blob/replica/transfer receipt.

With the transfer invocation as an fx of replica/allocate will the described polling work?

Reply:
yes, the blob/replica/transfer receipt will become available to the client when `ucan/conclude` is invoked with said receipt on the upload service at the end of replication